### ISSUE TITLE
Version manifesto: Tupling version part 2

### ISSUE THREAD

**Issue by trws** [LLNL/spack#30545]
In #22973 I reworked `Version` to do its comparisons directly on tuples representing the version rather than in python code with special cases.  At the time it was about a 2x perf increase on both construction and comparisons.  Since then, commit versions came in as well as a number of other new special cases.  There are also some new special cases that were a bit odd, #30544 implements phase one of this plan to move a lot of the extra logic out of the hot path.  Even with that however, coercion of versions and other version-related operations completely dominate the profile of a solve, with 6 of the top 8 functions (by own time) in a yappi trace being functions in `version.py`

![image](https://user-images.githubusercontent.com/660149/167229227-ecd6bdfa-5cf2-4f25-9cfc-19ecb3fc7b30.png)

I looked into this, and came to the following conclusions:

* The getitem cost is all lost time from using `self[slice]` in `satisfies` to generate a string of the version and create a new version from that string.  This one's easy to squash, PR forthcoming [ ]
* Initializing Version and VersionRange isn't very expensive, but it happens so much it needs to get even more efficient.  The generator expression inside of Version's constructor is actually the most called function in spack, and it probably shouldn't exist at all.  Looking into this, caching version objects given a specific string version may be a good solution, interned strings from packages should make that fast. [ ]
* Coercing versions is really, really useful for specializing logic for each kind of version, but we compare versions *a lot* and in the end `coercing_method` came out as the single most expensive function in spack for a solve.  So, can we get rid of this? I think so, how below. [ ]

Version representation to get rid of coercion for comparisons (at least, still thinking through other operations) is a triply nested tuple, roughly analagous to a versionlist containing a versionrange containing two versions.  Here are some examples to show what I mean:

```python
In [1]: import spack.compiler # import cycle in version, *sigh* always things to fix

In [2]: import spack.version as v

In [3]: v.Version("1.0")._cmp_version
Out[3]: (((1, 0), (1, 0, inf)),)

In [4]: v.Version("1.1")._cmp_version
Out[4]: (((1, 1), (1, 1, inf)),)

In [5]: v.Version("meh")._cmp_version
Out[5]: (((meh,), (meh, inf)),)

In [6]: v.VersionRange("1.0","1.0")._cmp_version
Out[6]: (((1, 0), (1, 0, inf)),)

In [7]: v.VersionRange("1.0","1.1")._cmp_version
Out[7]: (((1, 0), (1, 1, inf)),)

In [8]: v.VersionRange("1.0",None)._cmp_version
Out[8]: (((1, 0), (-inf,)),)

In [9]: v.VersionRange(None,None)._cmp_version
Out[9]: (((-inf,), (-inf,)),)

In [10]: v.VersionList(["1.0"])._cmp_version
Out[10]: (((1, 0), (1, 0, inf)),)
```

A few important things to note:
* The representations of `Version('1.0')`, `VersionRange('1.0','1.0')` and `VersionList(['1.0'])` are all *identical*.  So for ordering, versions, versionranges and versionlists can all be compared without coercion.
* There are now `float` components in the comparison version.  The idea is that we have `VersionStrComponent` to handle correct ordering of string versions when they are used, but we need to be able to represent two special cases:
  * Version upper bounds, logically are `<version>.inf`
  * VersionRange start and end are both open-ended if `None`.  
* So they are logically equivalent to `-inf` and `inf` respectively for comparisons to ints.  `VersionStrComponent` has to special case them, but it has to special case `int` too so that doesn't mean much.
* There are never any `None` values inside the comparison version tuple

End result, there is a stable comparison representation, unless there's a string version in it the comparison happens entirely in C code.  The string versions could probably be represented as well, but the resulting tuple would be a lot harder to parse for a human than these (in my opinion these aren't too bad).  One performance downside is that this means we have to convert the list inside a VersionList to a tuple for comparison.  Reworking it to only do that when a set of operations is done rather than on every add makes that pretty cheap though.

**Comment by trws** [comment]
One case I just realized may cause an issue. Not with the comparison versions specifically but with commit versions in general. When the commit version resolves, assuming it resolves to a version, its position in any VersionList that contains it needs to change. I'm not sure if we have actual bugs because of this, but because VersionList maintains everything by using an insertion sort on a list, it happens, and probably doesn't get fixed unless the usage happens to also re-sort the list. @becker33 I'm considering propagating the method to VersionList so spec can call it there and re-sort when complete, that would help with efficiently updating the comparison tuple anyway. That work for you?

**Comment by becker33** [comment]
@trws if we take #30544 and modify the `generate_commit_lookup` method to reorder based on the new tuples, is that sufficient for the issue or do you think it requires more than that?

The other issue is that if I understand this correctly, it changes the semantics for some comparisons. 
```
>>> from spack.version import ver
>>> (((1, 0), (1, 1, float('inf'))),) < (((1, 0), (-float('inf'),)),)
False
>>> (((1, 0), (1, 1, float('inf'))),) < (((1, 0), (float('inf'),)),)
True
>>> ver('1.0:1.1') < ver('1.0:')
True
```
I think the translation needs to be modified so that `None` as the start of a range is `(-inf,)` and `None` as the end of a range is `(inf,)`

**Comment by trws** [comment]
Yes, I think that's sufficient.  I reworked the generate_commit_lookup to chain down into inner versions already, but it needs the reordering fix.  The tuple version comparisons do need to use -inf and inf for none at start and end, in fact I think my local prototype currently does that, the one above was buggy and had them reversed, good catch!
